File: frontend/src/features/library/utils/subtitleStylePresets.ts
import type {
  AssStyleSpecDTO,
  LibraryBilingualLayoutDTO,
  LibraryBilingualStyleDTO,
  LibraryModuleConfigDTO,
  LibraryMonoStyleDTO,
  LibraryMonoStyleSnapshotDTO,
} from "@/shared/contracts/library"
import { t } from "@/shared/i18n"

export type SubtitleStylePresetKind = "mono" | "bilingual"

export type SubtitleStylePresetSelection =
  | {
      kind: SubtitleStylePresetKind
      id: string
    }
  | null

export const SUBTITLE_STYLE_ASPECT_RATIO_OPTIONS = [
  { value: "16:9", label: "16:9 · 1920×1080" },
  { value: "16:10", label: "16:10 · 1920×1200" },
  { value: "4:3", label: "4:3 · 1440×1080" },
  { value: "1:1", label: "1:1 · 1080×1080" },
  { value: "9:16", label: "9:16 · 1080×1920" },
] as const

export const BILINGUAL_BLOCK_ANCHOR_OPTIONS = [
  { value: 1, label: "Bottom Left" },
  { value: 2, label: "Bottom Center" },
  { value: 3, label: "Bottom Right" },
  { value: 4, label: "Middle Left" },
  { value: 5, label: "Middle Center" },
  { value: 6, label: "Middle Right" },
  { value: 7, label: "Top Left" },
  { value: 8, label: "Top Center" },
  { value: 9, label: "Top Right" },
] as const

export function resolveMonoStyles(config?: LibraryModuleConfigDTO) {
  return config?.subtitleStyles?.monoStyles ?? []
}

export function resolveBilingualStyles(config?: LibraryModuleConfigDTO) {
  return config?.subtitleStyles?.bilingualStyles ?? []
}

export function createEmptyAssStyleSpec(): AssStyleSpecDTO {
  return {
    fontname: "Arial",
    fontsize: 48,
    primaryColour: "&H00FFFFFF",
    secondaryColour: "&H00FFFFFF",
    outlineColour: "&H00111111",
    backColour: "&HFF111111",
    bold: false,
    italic: false,
    underline: false,
    strikeOut: false,
    scaleX: 100,
    scaleY: 100,
    spacing: 0,
    angle: 0,
    borderStyle: 1,
    outline: 2,
    shadow: 0.8,
    alignment: 2,
    marginL: 72,
    marginR: 72,
    marginV: 56,
    encoding: 1,
  }
}

export function createEmptyMonoStyle(): LibraryMonoStyleDTO {
  return {
    id: createSubtitleStylePresetID("mono"),
    name: t("library.config.subtitleStyles.newMonoStyle"),
    builtIn: false,
    basePlayResX: 1920,
    basePlayResY: 1080,
    baseAspectRatio: "16:9",
    style: createEmptyAssStyleSpec(),
  }
}

export function createMonoSnapshotFromStyle(style: LibraryMonoStyleDTO): LibraryMonoStyleSnapshotDTO {
  return {
    sourceMonoStyleID: style.id,
    sourceMonoStyleName: style.name,
    name: style.name,
    basePlayResX: style.basePlayResX,
    basePlayResY: style.basePlayResY,
    baseAspectRatio: style.baseAspectRatio,
    style: cloneAssStyleSpec(style.style),
  }
}

export function createEmptyBilingualLayout(): LibraryBilingualLayoutDTO {
  return {
    gap: 24,
    blockAnchor: 2,
  }
}

export function createEmptyBilingualStyle(monoStyles: LibraryMonoStyleDTO[] = []): LibraryBilingualStyleDTO {
  const primarySource = monoStyles[0] ?? createEmptyMonoStyle()
  const secondarySource = monoStyles[1] ?? monoStyles[0] ?? createEmptyMonoStyle()
  return {
    id: createSubtitleStylePresetID("bilingual"),
    name: t("library.config.subtitleStyles.newBilingualStyle"),
    builtIn: false,
    basePlayResX: primarySource.basePlayResX,
    basePlayResY: primarySource.basePlayResY,
    baseAspectRatio: primarySource.baseAspectRatio,
    primary: createMonoSnapshotFromStyle(primarySource),
    secondary: createMonoSnapshotFromStyle(secondarySource),
    layout: createEmptyBilingualLayout(),
  }
}

export function cloneAssStyleSpec(style: AssStyleSpecDTO): AssStyleSpecDTO {
  return { ...style }
}

export function cloneMonoStyle(style: LibraryMonoStyleDTO): LibraryMonoStyleDTO {
  return {
    ...style,
    style: cloneAssStyleSpec(style.style),
  }
}

export function cloneMonoStyleSnapshot(style: LibraryMonoStyleSnapshotDTO): LibraryMonoStyleSnapshotDTO {
  return {
    ...style,
    style: cloneAssStyleSpec(style.style),
  }
}

export function cloneBilingualStyle(style: LibraryBilingualStyleDTO): LibraryBilingualStyleDTO {
  return {
    ...style,
    primary: cloneMonoStyleSnapshot(style.primary),
    secondary: cloneMonoStyleSnapshot(style.secondary),
    layout: { ...style.layout },
  }
}

export function applyAspectRatioBaseResolution(aspectRatio: string) {
  switch (aspectRatio) {
    case "16:10":
      return { basePlayResX: 1920, basePlayResY: 1200 }
    case "4:3":
      return { basePlayResX: 1440, basePlayResY: 1080 }
    case "1:1":
      return { basePlayResX: 1080, basePlayResY: 1080 }
    case "9:16":
      return { basePlayResX: 1080, basePlayResY: 1920 }
    default:
      return { basePlayResX: 1920, basePlayResY: 1080 }
  }
}

export function createSubtitleStylePresetID(prefix: SubtitleStylePresetKind) {
  return `${prefix}-${Date.now().toString(36)}-${Math.random().toString(36).slice(2, 8)}`
}

File: internal/application/library/dto/subtitle_style_presets.go
package dto

import "encoding/json"

type AssStyleSpecDTO struct {
	Fontname           string  `json:"fontname"`
	FontFace           string  `json:"fontFace,omitempty"`
	FontWeight         int     `json:"fontWeight,omitempty"`
	FontPostScriptName string  `json:"fontPostScriptName,omitempty"`
	Fontsize           float64 `json:"fontsize"`
	PrimaryColour      string  `json:"primaryColour"`
	SecondaryColour    string  `json:"secondaryColour"`
	OutlineColour      string  `json:"outlineColour"`
	BackColour         string  `json:"backColour"`
	Bold               bool    `json:"bold"`
	Italic             bool    `json:"italic"`
	Underline          bool    `json:"underline"`
	StrikeOut          bool    `json:"strikeOut"`
	ScaleX             float64 `json:"scaleX"`
	ScaleY             float64 `json:"scaleY"`
	Spacing            float64 `json:"spacing"`
	Angle              float64 `json:"angle"`
	BorderStyle        int     `json:"borderStyle"`
	Outline            float64 `json:"outline"`
	Shadow             float64 `json:"shadow"`
	Alignment          int     `json:"alignment"`
	MarginL            int     `json:"marginL"`
	MarginR            int     `json:"marginR"`
	MarginV            int     `json:"marginV"`
	Encoding           int     `json:"encoding"`
}

type LibraryMonoStyleDTO struct {
	ID                 string          `json:"id"`
	Name               string          `json:"name"`
	BuiltIn            bool            `json:"builtIn,omitempty"`
	BasePlayResX       int             `json:"basePlayResX"`
	BasePlayResY       int             `json:"basePlayResY"`
	BaseAspectRatio    string          `json:"baseAspectRatio"`
	SourceAssStyleName string          `json:"sourceAssStyleName,omitempty"`
	Style              AssStyleSpecDTO `json:"style"`
}

type LibraryMonoStyleSnapshotDTO struct {
	SourceMonoStyleID   string          `json:"sourceMonoStyleID,omitempty"`
	SourceMonoStyleName string          `json:"sourceMonoStyleName,omitempty"`
	Name                string          `json:"name"`
	BasePlayResX        int             `json:"basePlayResX"`
	BasePlayResY        int             `json:"basePlayResY"`
	BaseAspectRatio     string          `json:"baseAspectRatio"`
	Style               AssStyleSpecDTO `json:"style"`
}

type LibraryBilingualLayoutDTO struct {
	Gap         float64 `json:"gap"`
	BlockAnchor int     `json:"blockAnchor"`
}

type LibraryBilingualStyleDTO struct {
	ID              string                      `json:"id"`
	Name            string                      `json:"name"`
	BuiltIn         bool                        `json:"builtIn,omitempty"`
	BasePlayResX    int                         `json:"basePlayResX"`
	BasePlayResY    int                         `json:"basePlayResY"`
	BaseAspectRatio string                      `json:"baseAspectRatio"`
	Primary         LibraryMonoStyleSnapshotDTO `json:"primary"`
	Secondary       LibraryMonoStyleSnapshotDTO `json:"secondary"`
	Layout          LibraryBilingualLayoutDTO   `json:"layout"`
}

type DCSSPFileDTO struct {
	Format        string          `json:"format"`
	SchemaVersion int             `json:"schemaVersion"`
	Type          string          `json:"type"`
	ID            string          `json:"id"`
	Name          string          `json:"name"`
	Author        string          `json:"author,omitempty"`
	Homepage      string          `json:"homepage,omitempty"`
	Description   string          `json:"description,omitempty"`
	License       string          `json:"license,omitempty"`
	Tags          []string        `json:"tags,omitempty"`
	CreatedAt     string          `json:"createdAt,omitempty"`
	UpdatedAt     string          `json:"updatedAt,omitempty"`
	AppVersion    string          `json:"appVersion,omitempty"`
	Payload       json.RawMessage `json:"payload"`
}

type DCSSPMonoPayloadDTO struct {
	BasePlayResX    int             `json:"basePlayResX"`
	BasePlayResY    int             `json:"basePlayResY"`
	BaseAspectRatio string          `json:"baseAspectRatio"`
	Style           AssStyleSpecDTO `json:"style"`
}

type DCSSPBilingualPayloadDTO struct {
	BasePlayResX    int                         `json:"basePlayResX"`
	BasePlayResY    int                         `json:"basePlayResY"`
	BaseAspectRatio string                      `json:"baseAspectRatio"`
	Primary         LibraryMonoStyleSnapshotDTO `json:"primary"`
	Secondary       LibraryMonoStyleSnapshotDTO `json:"secondary"`
	Layout          LibraryBilingualLayoutDTO   `json:"layout"`
}

type GenerateSubtitleStylePreviewASSRequest struct {
	Type          string                        `json:"type"`
	Mono          *LibraryMonoStyleDTO          `json:"mono,omitempty"`
	Bilingual     *LibraryBilingualStyleDTO     `json:"bilingual,omitempty"`
	FontMappings  []LibrarySubtitleStyleFontDTO `json:"fontMappings,omitempty"`
	PrimaryText   string                        `json:"primaryText,omitempty"`
	SecondaryText string                        `json:"secondaryText,omitempty"`
}

type GenerateSubtitleStylePreviewASSResult struct {
	ASSContent             string   `json:"assContent"`
	ReferencedFontFamilies []string `json:"referencedFontFamilies,omitempty"`
}

type GenerateSubtitleStylePreviewVTTRequest struct {
	Type          string                        `json:"type"`
	Mono          *LibraryMonoStyleDTO          `json:"mono,omitempty"`
	Bilingual     *LibraryBilingualStyleDTO     `json:"bilingual,omitempty"`
	FontMappings  []LibrarySubtitleStyleFontDTO `json:"fontMappings,omitempty"`
	PrimaryText   string                        `json:"primaryText,omitempty"`
	SecondaryText string                        `json:"secondaryText,omitempty"`
	PreviewWidth  int                           `json:"previewWidth,omitempty"`
	PreviewHeight int                           `json:"previewHeight,omitempty"`
}

type GenerateSubtitleStylePreviewVTTResult struct {
	VTTContent string `json:"vttContent"`
}

type GenerateWorkspacePreviewASSRequest struct {
	LibraryID                string `json:"libraryId"`
	DisplayMode              string `json:"displayMode,omitempty"`
	PrimarySubtitleTrackID   string `json:"primarySubtitleTrackId"`
	SecondarySubtitleTrackID string `json:"secondarySubtitleTrackId,omitempty"`
}

type GenerateWorkspacePreviewASSResult struct {
	ASSContent             string   `json:"assContent"`
	ReferencedFontFamilies []string `json:"referencedFontFamilies,omitempty"`
}

type ParseSubtitleStyleImportRequest struct {
	Content  string `json:"content"`
	Format   string `json:"format,omitempty"`
	Filename string `json:"filename,omitempty"`
}

type ParseSubtitleStyleImportResult struct {
	ImportFormat       string                    `json:"importFormat"`
	DCSSP              *DCSSPFileDTO             `json:"dcssp,omitempty"`
	MonoStyles         []LibraryMonoStyleDTO     `json:"monoStyles,omitempty"`
	BilingualStyle     *LibraryBilingualStyleDTO `json:"bilingualStyle,omitempty"`
	DetectedRatio      string                    `json:"detectedRatio,omitempty"`
	NormalizedPlayResX int                       `json:"normalizedPlayResX,omitempty"`
	NormalizedPlayResY int                       `json:"normalizedPlayResY,omitempty"`
	Warnings           []string                  `json:"warnings,omitempty"`
}

type ExportSubtitleStylePresetRequest struct {
	DirectoryPath string                    `json:"directoryPath"`
	Type          string                    `json:"type"`
	Mono          *LibraryMonoStyleDTO      `json:"mono,omitempty"`
	Bilingual     *LibraryBilingualStyleDTO `json:"bilingual,omitempty"`
}

type ExportSubtitleStylePresetResult struct {
	ExportPath string `json:"exportPath"`
	FileName   string `json:"fileName"`
}
